은행 창구 관리 앱 [Step3] 록스, 재닌

janine-kang opened the pull request:
리뷰어: @jungseungyeo
버드: @janine-kang, @KimRoks

린생, 안녕하세요!
아래와 같이 STEP3의 PR 요청드립니다! 🙌🏻

이번에도 잘부탁드립니다 🔥

## Summary
STEP3 OperationQueue와 BlockOperation을 이용한 동시성 프로그램 구현

![STEP3](https://github.com/tasty-code/ios-bank-manager/assets/76927263/528596a0-32e7-45ed-aca1-3d206e717940)


### 구현 사항
- [x] GCD를 활용한 동시성 프로그래밍 구현
- [x] 스레드(Thread) 개념에 대한 이해
- [x] 동시성 프로그래밍을 위한 기반기술(GCD, Operation) 등의 이해 
- [x] 동시성 프로그래밍 개념의 이해
- [x] 동기(Synchronous)와 비동기(Asynchronous)의 이해
- [x] 동기(Synchronous)와 비동기(Asynchronous) 동작의 구현 및 적용

### 주요 고려 사항
- 업무를 진행하는 은행원을 `Thread`로 이해하고, 추상화에서 제외
- 예금 창구와 대출 창구를 구현하기 위해 두개의 `OperationQueue`로 분리
- 각각의 창구 개수를 위해 `maxConcurrentOperationCount`를 이용하여 동시 작업 수 제한
- `waitUntilAllOperationsAreFinished` 이용하여 전체 작업 완료 시점 파악
- 재사용성 확보를 위해 OperationQueue의 `convenience initializer` 정의
- 고객의 업무 유형 `ServiceType` 모델 추가

### 궁금한 점
1. 일전에 린생께서 말씀하셨던 `bank.open(with: Int)` 메서드에 대해서 고민해보았는데요. 

open()에서 고객의 수를 받는 궁극적인 이유는 대기열에 enqueue하기 위함이었고, 그러므로 복수 개의 Customer를 생성하여 `waitingLine` 프로퍼티에 enqueue하는 방식으로 되어있었는데요. 이 과정에서 복수의 Customer instance를 넘기게 된다면 또다른 Collection을 사용해야 할 것이라고 판단하여 bank.open()의 Parameter로 Customer Model을 넘기기엔 무리라는 판단을 하게 되었습니다. 

그렇다고 waitingLine 자체의 소유권을 BankManager가 가지고 있다가 이 자체를 Bank에게 던져주게 된다면, 추후 고객이 추가적으로 늘어나는 경우를 대비할 수 없어 확장성이 닫히게 되므로 좋지 않은 판단이라고 생각하였으며, 
waitingLine 자체에 대한 접근 제어를 낮추게 된다면 `dequeue` 작업에 대한 은닉화가 저해되므로 이 또한 피해야 한다고 생각하게 되었습니다.

따라서 `enqueue` 작업을 하는 메서드 자체를 Bank가 가지고 있고, 그에 대한 접근제어를 internal로 약화하여 외부에서 접근할 수 있는 쪽으로 가닥을 잡았습니다. 
그 결과는 아래와 같습니다.

```swift
final class Bank {
    func runService() {
        startService()
        shutDownService()
    }
    
    func lineUp(_ customer: Customer) { 
        waitingLine.enqueue(customer)
    }
}
```
```swift
extension BankManager {
    func enqueueCustomers(upTo numberOfCustomers: Int) {
        for i in numberOfCurrentCustomers + 1...numberOfCurrentCustomers + initialCustomers {
            bank.lineUp(Customer(ticketNumber: i))
        }
        
        numberOfCurrentCustomers += numberOfCustomers
    }
}
```
저희가 고민했던 방법 이외에 린생이 가지고 계셨던 방법이 분명히 있을거라고 생각하는데요. 그 부분에 대해서 공유해주시면 성장에 많은 도움이 될 것 같습니다 🙇🏻‍♀️

2. 이 고민은 OperationQueue를 채택하고나서 부딪혔던 문제에 관련된 것인데요. 대출 업무와 예금 업무 각각이 동시에 실행될 수 있는 수가 제한되어 있어 생긴 궁금증입니다. 
```
 func startService() {
        let loanQueue: OperationQueue = OperationQueue(name: "LoanQueue", maxConcurrentOperationCount: 1)
        let depositQueue: OperationQueue = OperationQueue(name: "DepositQueue", maxConcurrentOperationCount: 2)
        
        while !waitingLine.isEmpty {
            guard let currentCustomer = waitingLine.dequeue() else { break }
            
            let taskBlock = BlockOperation {
                self.provideService(to: currentCustomer)
            }
            
            if currentCustomer.serviceType == .deposit {
                depositQueue.addOperation(taskBlock)
            } else {
                loanQueue.addOperation(taskBlock)
            }
        }
        
        OperationQueue.waitUntilAllOperationsAreFinished(depositQueue, loanQueue)
    }
```
현재는 OperationQueue 위와 같이 자체를 두 개로 분리하여 진행하고 있는데, 혹시 이 부분을 하나의 OperationQueue를 이용하여 구현될 수 있는 방법이 있을까요? 검색해봤을 때 이렇다 할 방법이 나오지 않았는데, 혹시 있다면 알고싶습니다!

바쁘실텐데 꼼꼼하게 리뷰해주셔서 감사드립니다! 이번 STEP도 잘 부탁드립니다 린생 🙇🏻‍♀️

janine-kang committed:
feat: create ServiceType

KimRoks committed:
feat: expend customer type

janine-kang committed:
feat: expand proviceService with serviceType

janine-kang committed:
feat: fix print message in provideService()

KimRoks committed:
fix: modify GCD to Operation

janine-kang committed:
fix: modify durationTime Type

janine-kang committed:
feat: add OperationQueue for dividing worker

KimRoks committed:
refactor: delete default case

- BankManager의 launch() default case 삭제조치

janine-kang committed:
refactor: modify method
- Bank 타입 관심사 분리

janine-kang committed:
Merge pull request #1 from KimRoks/feat/STEP3

STEP3 PR

janine-kang committed:
feat: add resetCustomerCount()

KimRoks committed:
Merge branch 'feat/STEP3' into Develop

janine-kang committed:
fix: modify vnqueueCustomers()
- initialCustomers 값 캡처

janine-kang committed:
Merge pull request #2 from KimRoks/feat/STEP3

fix: modify enqueueCustomers()

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Controllers/BankManager.swift:
```diff
@@ -2,26 +2,38 @@ import Foundation
         print(Prompt.appLaunch, terminator: " ")
-        if let userInput = readLine() {
-            switch Menu(rawValue: userInput) {
+        if let userInput = readLine(), let menu = Menu(rawValue: userInput) {
```
수정이 되었네요 ㅎㅎ 너무 좋습니다 ㅎㅎ

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Extensions/OperationQueue.swift:
```diff
@@ -0,0 +1,21 @@
+        queue.forEach { $0.waitUntilAllOperationsAreFinished() }
+    }
+    
+    convenience init(name: String, maxConcurrentOperationCount: Int?) {
```
확장해서 초기화하는 함수를 만들었네요 너무 좋습니다.

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Controllers/BankManager.swift:
```diff
@@ -2,26 +2,38 @@ import Foundation
-
+private extension BankManager {
+    func enqueueCustomers(upTo numberOfCustomers: Int) {
+        for i in numberOfCurrentCustomers + 1...numberOfCurrentCustomers + numberOfCustomers {
```
```suggestion
        for i in 1 ... numberOfCustomers {
```

jungseungyeo replied:
위와 같이 수정해도 될거같은데 이렇게 작성된 의도가 있나요?

janine-kang replied:
추후 고객이 추가될 경우를 대비해서 이렇게 만들게 되었는데요.
고객 객체 안에 있는 ticketNumber가 이 for-loop을 돌면서 부여되기 때문에 위처럼 생성하게 되었습니다!

사실 overloading을 통해 두 개의 경우로 나눌까도 고민했는데, 이렇게 offset도 동적으로 할당하면 하나의 메서드를 재사용할 수 있다고 판단해서 위 처럼 만들게되었습니다!

jungseungyeo replied:
그런 의도였다면 하나의 변수로 명확한 네이밍으로 변환 후 작업하면 좋을거같아요. 코드상으로 왜 + 가 되어야 하는지 명확한 이유를 알지못해서 휴먼 에러가 발생 할 수있는 요소가 보여요

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -9,62 +9,60 @@ import Foundation
+private extension Bank {
+    func startService() {
+        let loanQueue: OperationQueue = OperationQueue(name: "LoanQueue", maxConcurrentOperationCount: 1)
+        let depositQueue: OperationQueue = OperationQueue(name: "DepositQueue", maxConcurrentOperationCount: 2)
```
OperationQueue를 사용하셨군요 ㅎㅎ OperationQueue는 어떤 기능이 있을까요?

KimRoks replied:
OperationQueue는 기본적으로 isAsynchronous == true 로 non - concurrent 작업을 넣어도 concurrent하게 작동하고 종속성을 추가 할 수 있는 등 여러 기능이 있지만
그 중 maxConcurrentOperationCount()를 사용하기 위한 선택이었습니다 실질적인 쓰레드의 개수를 설정해줄 수 있는 메서드라고 판단했기때문에 OperationQueue를 선택하여 사용하게되었습니다

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -9,62 +9,60 @@ import Foundation
+        let durationTime: UInt32 = UInt32(serviceType.duration * 1_000_000)
+        
+        print(Prompt.serviceStart(customer: target.ticketNumber, service: serviceType.description))
+        usleep(durationTime)
```
usleep을 하면 무슨 일이 발생할까요?

KimRoks replied:
usleep은 현재 프로세스의 실행을 지연시키는데 사용됩니다
일반 sleep과 다른 부분은 sleep은 초단위로 작동하기에 이번 구현 사항과는 맞지않다 판단하였고
usleep의 단위는 마이크로초(1/1,000,000 초)로 소수처리까지 해야하는 이번 프로젝트에 적합하다고 판단하였습니다

jungseungyeo replied:
주로 sleep을 많이 사용하지는 않아요 sleep을 사용하면 이야기 한거처럼 스레드를 멈추게 되어 비동기 작업 시 어떤 스레드가 동작을 멈추어 오작동하는지 파악하기도 어려워져요. 이런경우 주로 GCD의 Dispatchqueue의 asyncAfter를 사용하는 편입니다. 참고해주세요 :)

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/ServiceType.swift:
```diff
@@ -0,0 +1,31 @@
+
+enum ServiceType: CustomStringConvertible, CaseIterable {
+    case loan
+    case deposit
```
enum으로 정의되어서 관리하기가 좋네요 ㅎㅎ

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -9,62 +9,60 @@ import Foundation
-                self?.provideService(to: frontCustomer)
+            let taskBlock = BlockOperation {
+                self.provideService(to: currentCustomer)
+            }
```
DispatchQueue를 사용하다가 Operation으로 변경을 하셨네요 ㅎㅎ 두 차이점이 무엇인지 아시나요?

KimRoks replied:
Dispatch는 GCD의 객체로 동시성 프로그래밍을 위한 API입니다
operation은 GCD의 task를  객체지향적으로 풀어낸 API로 캡슐화, 객체화(타입화)되었기 때문에 객체지향이 갖는 장점과 유사한 장점들을 가집니다

jungseungyeo replied:
async/await 영상이긴하지만 GCD 내용이 있어 이 영상도 참고해보세요 ㅎㅎ

https://developer.apple.com/videos/play/wwdc2021/10254/

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -9,62 +9,60 @@ import Foundation
+    func shutDownService() {
         print(Prompt.appFinish(totalCustomer: exitCount, totalTime: totalTime))
-        clear()
+        prepareService()
```
clear보다 네이밍이 더 직관적이여서 좋네요

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -9,62 +9,60 @@ import Foundation
+            if currentCustomer.serviceType == .deposit {
+                depositQueue.addOperation(taskBlock)
+            } else {
+                loanQueue.addOperation(taskBlock)
```
여기서 분기문이 들어가는 부분이 아쉽네요.. 흠.. 작업이 많아지면 분기문이 늘어날수있는 요소가 보여요. 
현재는 큰 문제가 없겠지만 serviceType을 알지 못하여 queue에 추상화되어서 들어갈수있는구조로 개선이 되면 더욱좋을거같네요.
흠.. 그런데 그런 구조가 딱히 생각이 나지 않네요 ㅎㅎ.. 이부분은 인지하고 넘어가시면 좋을거같네요.

janine-kang replied:
https://github.com/tasty-code/ios-bank-manager/pull/46/files#r1387404386

저희도 이 부분에 대해서 많은 고민을 해보았고, 결론적으로는 의존성 문제와 OCP 확보를 위해 `OperationQueue` -> `DispatchQueue + DispatchGroup + DispatchSemaphore`의 조합으로 변경하고,
BankServiceExecutor라는 Type을 추가하였습니다.

저희가 처음 Operation으로 넘어갔던 이유가 1) 동시에 돌아갈 수 있는 스레드 수 제한 2) 전체 작업이 끝날때 까지 wait() 시키기라는 두 작업을 위해서였는데요. 
예금업무와 대출업무를 부득이하게 2개의 Queue로 나누어야 한다는 점과, 동시에 2개의 Queue로 나눠지면서 각각 Queue에 부합하는 업무를 따로 던져줘야 하므로 확장성을 낮춘다는 점을 개선하고 싶었습니다.

그 결과로 하나의 DispatchQueue를 이용하여 한개의 Queue로 통일하고, 이를 Group으로 묶어 wait()에 대한 작업도 확보할 수 있었습니다. DispatchSemaphore를 이용하여 예금 업무와 대출 업무의 동시 진행 가능 수를 제어하고, BankServiceExecutor를 이용하여 ServiceType이 추가되어도 따로 변경하지 않게끔 코드를 수정했습니다. 이 부분에 대해서 린생이 어떻게 보시는지 궁금합니다 👀

jungseungyeo replied:
Operation로 수정을 다시 ㅎㅎ.. 뭔가 오해가 있는거 같아요. 변경 된 코드를 보면 thread가 고객수 만큼 생성이 됩니다. 반복문을 돌면서 스래드를 생성하게 되게 수정이 된거같아요. 일단 머지를 할테니 Operation를 사용하는 방법으로 수정부탁드립니다.

jungseungyeo commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -9,62 +9,60 @@ import Foundation
+        usleep(durationTime)
+        print(Prompt.serviceDone(customer: target.ticketNumber, service: serviceType.description))
+        
+        totalTime += serviceType.duration
```
이런식으로 총 시간을 더해지면 업무시간이 아니라 모든 직원의 근무시간인거같습니다. ㅠ

jungseungyeo replied:
예금 + 예금 = 0.7, 대출 = 1.1 이렇게 동시 수행되는데 위의 totaltime을 += 처리 되면 3.5초로 업무시간이 측정되는거같네요

KimRoks replied:
DispatchTime.now()을 사용하여 수정 완료하였습니다
https://github.com/tasty-code/ios-bank-manager/pull/46/files#r1387412739

jungseungyeo reviewed (changes requested):
1.일전에 린생께서 말씀하셨던 bank.open(with: Int) 메서드에 대해서 고민해보았는데요.

`저도 작성해주신 코드와 유사하게 생각을 하였습니다. 저만의 특별한 방법이 있을거라고 생각한거같으나.. 그런게 있지는 않고 단순히 고객 모델로 정보를 담아서 Bank가 처리하는 구조를 생각하였습니다. 
작성해주신 코드를 보면 제가 의도 한 대로 잘 적용이 되었고, 업무 정보 또한 잘 적용 된거같습니다 
`

2. 이 고민은 OperationQueue를 채택하고나서 부딪혔던 문제에 관련된 것인데요. 대출 업무와 예금 업무 각각이 동시에 실행될 수 있는 수가 제한되어 있어 생긴 궁금증입니다.

`생각을 해보다가 depositQueue 를 maxConcurrentOperationCount  3으로 두고 하나로 사용해도 문제가 없어보여요. 이렇게 사용하면 serviceTtype 도 체크 안해도 되고 좋을거같아요. maxConcurrentOperationCount 1로 수행하면 serial queue처럼 동작이 되지만 2개가 되면 serial queue이 아니게 되죠`

전반적으로 구조가 좋은거같아요 ㅎㅎ 동작의 흐름도 잘보이고 OperationQueue를 확장하여 잘 사용한거같습니다. 간단하게 코멘트 작성과 OperationQueue를 하나로 적용하고 업무시간 측정을 요구사항에 맞게 수정 부탁드려요 :)

janine-kang committed:
refactor: modify structure
- Operation에서 DispatchQueue로 방식 변경
- DispatchSemaphore를 이용하여 현재 진행되는 작업 수 제어
- DispatchSemaphore 이용하여 Race Condition 제어
- BankServiceExecutor로 OCP 확보와 의존성 주입
- totalTime을 기존 모든 직원의 근무시간의 합에서 총 걸린 시간으로 변경

KimRoks committed:
refactor: rename properties and methods

- 더 나은 가독성을 위한 리네임 작업
- checkExecutedTime(start: , end: ) -> calculateTotalWorkTime(from :, to: )
- totalTime -> totalWorkTime

KimRoks committed:
Merge branch 'feat/STEP3' into Develop

janine-kang commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/BankServiceExecutor.swift:
`BankServiceExecutor`를 추가한 부분입니다

janine-kang commented:
> 1.일전에 린생께서 말씀하셨던 bank.open(with: Int) 메서드에 대해서 고민해보았는데요.
> 
> `저도 작성해주신 코드와 유사하게 생각을 하였습니다. 저만의 특별한 방법이 있을거라고 생각한거같으나.. 그런게 있지는 않고 단순히 고객 모델로 정보를 담아서 Bank가 처리하는 구조를 생각하였습니다. 작성해주신 코드를 보면 제가 의도 한 대로 잘 적용이 되었고, 업무 정보 또한 잘 적용 된거같습니다 `
> 
> 2. 이 고민은 OperationQueue를 채택하고나서 부딪혔던 문제에 관련된 것인데요. 대출 업무와 예금 업무 각각이 동시에 실행될 수 있는 수가 제한되어 있어 생긴 궁금증입니다.
> 
> `생각을 해보다가 depositQueue 를 maxConcurrentOperationCount 3으로 두고 하나로 사용해도 문제가 없어보여요. 이렇게 사용하면 serviceTtype 도 체크 안해도 되고 좋을거같아요. maxConcurrentOperationCount 1로 수행하면 serial queue처럼 동작이 되지만 2개가 되면 serial queue이 아니게 되죠`
> 
> 전반적으로 구조가 좋은거같아요 ㅎㅎ 동작의 흐름도 잘보이고 OperationQueue를 확장하여 잘 사용한거같습니다. 간단하게 코멘트 작성과 OperationQueue를 하나로 적용하고 업무시간 측정을 요구사항에 맞게 수정 부탁드려요 :)

### 2번 피드백에 대하여
---
![image](https://github.com/tasty-code/ios-bank-manager/assets/76927263/929ed004-dfcd-43c5-847d-2775251a5037)

2번 부분에서, 저희가 2개로 나눈 이유는 예금이 2개, 대출이 1개로 동시에 처리될 수 있는 수가 제한되어있기 때문입니다.
처음에는 저희도 Queue 하나로 처리하고자 했었는데, 위의 제한 사항을 지키려고 하다보니 2개로 나눠지게 되었습니다.
이번 PR 받고 나서 구조적으로 조금 손을 보았는데요, 그러면서 이 부분도 함께 고쳐지게 되었습니다. 한번 확인해보시고 린생의 의견 전달해주시면 감사하겠습니다 🙇🏻‍♀️

janine-kang commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -8,63 +8,81 @@
-            guard let frontCustomer = waitingLine.dequeue() else { break }
+            guard let currentCustomer = waitingLine.dequeue() else { break }
             
-            queue.async(group: group) { [weak self] in
-                self?.provideService(to: frontCustomer)
+            DispatchQueue.global().async(group: group) { [weak self] in
+                guard let self = self, let worker = self.serviceList[currentCustomer.serviceType] else { return }
+                worker.work {
+                    self.provideService(to: currentCustomer)
+                }
             }
         }
-        
+
         group.wait()
-        close()
+        
+        let endTime = DispatchTime.now()
+        calculateTotalWorkTime(from: startTime, to: endTime)
```
이렇게 고쳐졌습니다!

jungseungyeo replied:
ㅠㅠ.. 이렇게 수정이 되어서 thread가 고객수 만큼 생성이 됩니다.  이거 Operation으로 사용하게 다시 수정 부탁드려요 ㅠ

KimRoks committed:
refactor: modify BankServiceExecutor

- class에서 struct로 변경

KimRoks committed:
Merge branch 'feat/STEP3' into Develop

jungseungyeo reviewed (changes requested):
request changes를 두번 요청드려서 죄송합니다.

Roks한테 디코에서 이야기 나누었습니다. 오퍼레이션 큐를 사용하는 방법으로 다시 수정해 부탁드려요 ㅎㅎ 
다른 부분은 다 좋습니다. ㅎㅎ 감사합니다.

janine-kang committed:
refactor: revert to OperationQueue

KimRoks committed:
Merge branch 'feat/STEP3' into Develop

janine-kang commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/BankServiceExecutor.swift:
```diff
@@ -8,15 +8,17 @@
     init(type: ServiceType) {
-        self.semaphore = DispatchSemaphore(value: type.maxCount)
+        self.queue = OperationQueue(name: type.description, maxConcurrentOperationCount: type.maxCount)
     }
     
-    func work(_ task: () -> Void) {
-        semaphore.wait()
-        task()
-        semaphore.signal()
+    func work(_ task: BlockOperation) {
+        queue.addOperation(task)
+    }
+    
+    func wait() {
+        queue.waitUntilAllOperationsAreFinished()
```
DispatchSemaphore -> OperationQueue로 변경했습니다.

janine-kang commented on BankManagerConsoleApp/BankManagerConsoleApp/Models/Bank.swift:
```diff
@@ -36,21 +36,19 @@ final class Bank {
+            guard let currentCustomer = waitingLine.dequeue(), let queue = self.serviceList[currentCustomer.serviceType] else { return }
             
-            DispatchQueue.global().async(group: group) { [weak self] in
-                guard let self = self, let worker = self.serviceList[currentCustomer.serviceType] else { return }
-                worker.work {
-                    self.provideService(to: currentCustomer)
-                }
+            let taskBlock = BlockOperation {
+                self.provideService(to: currentCustomer)
```
이 부분에서 다시 Operation Task로 감싸, currentCustomer의 유형에 맞는 Queue로 던져줍니다

janine-kang commented:
https://github.com/tasty-code/ios-bank-manager/pull/46/commits/c1892720b5f02a5694fcdb9ae73819b0384309dd#r1388793534

https://github.com/tasty-code/ios-bank-manager/pull/46/commits/c1892720b5f02a5694fcdb9ae73819b0384309dd#r1388794071

안녕하세요 린생!

린생께서 말씀해주신대로 OperationQueue로 작업 방식 변환했습니다.

다만 저희가 Thread에 대한 테스트도 진행해보았는데요.
OperationQueue에서 제공하는 maxConcurrentOperationCount라는 API 조차도 쓰레드를 제어하는 것이 아닌, 현재 동시에 진행될 수 있는 작업의 수를 받을 수 있게 만들어 준 것이므로, 항상 3개의 쓰레드만이 해당 서비스에 할당되었다고 보장할 수 없다는 것을 확인했습니다.

물론 이번 구현 방법이 기존의 방법인 하나의 DispatchQueue와 3개의 DispatchSemaphore를 통한 Task제어 방법으로 할 때에 비해 쓰레드의 할당이 불필요하게 많이 일어나지는 않습니다. 그러나 쓰레드가 무조건적으로 3개만 돌지는 않는다는 것을 인지할 수 있었습니다. 

이러한 이슈가 있음에도 명확하게 스레드의 수를 제어하지 않기로 결정한 이유는 추상화가 변경되었기 때문인데요. 
저희가 기존에는 쓰레드 자체를 은행원으로 추상화하여 따로 타입을 구현해내지 않은 채로, 쓰레드에게 일을 던져주는 것으로 은행원 역할을 대신했습니다. 
그러나 확장성을 보장하기 위해서 BankServiceExecutor 타입을 추가하면서 이 추상화가 변형되게 되었고, 현재는 실질적으로 동시에 처리할 수 있는 수를 제어할 수 있는 것만으로도 BankServiceExecutor가 은행원(엄밀히 말하면 각 부서인 것 같습니다)을 추상화한 존재라고 생각합니다.

따라서, 이 타입을 통해 operation queue를 생성하고 있으며, 이 방법이 가지고 있는 다른 잠재적인 위험성을 파악하진 못했으나 현재의 방법 내에서는 확장성을 보장할 수 있는 가장 최선의 방법이라고 생각합니다.

이 부분에 관련해서는 린생의 의견도 듣고 싶습니다. 코드를 보시고, 추상화 방면에서, 그리고 동시성에 대한 이슈에 관련해서도 코멘트 남겨주시면 감사하겠습니다.

jungseungyeo reviewed (approved):
janine, roks 고생하셨습니다.

요구조건을 잘 지켜서 만드셨습니다.

걱정해서 작성해주신 부분을 보면 큰 문제가 되지는 않는거같아요. OperationQueue를 하나로 만들고 사용하였고, 일을 추가해서 사용할때 enum(ServiceType)의 case 를 추가하면 확장이 가능하죠

이렇게 요구 조건과 확장성이 열려있다면 괜찮은 방법입니다. 
전의 방법도 크게 문제가 되지는 않지만 case가 추가 될 떄 마다 OperationQueue를 추가하여 사용해야 하는 단점이 있었죠. 이런 부분에서 많은 개선이 있는 코드로 보여 걱정해주는 부분의 보장이 되지 않더라도 이슈가 있을거같지는 않네요 ㅎㅎ

이번 스텝을 하면서 고생많으셨습니다. 다음 스텝으로 진행부탁드립니다.
